ewatercycle/models/marrmot.py

```diff
@@ -0,0 +1,213 @@
+import time
+from dataclasses import asdict, dataclass
+from datetime import datetime, timezone
+from os import PathLike
+from pathlib import Path
+from typing import Any, Iterable, Tuple, Union
+
+import numpy as np
+import scipy.io as sio
+import xarray as xr
+from cftime import num2date
+from grpc4bmi.bmi_client_docker import BmiClientDocker
+from grpc4bmi.bmi_client_singularity import BmiClientSingularity
+
+from ewatercycle import CFG
+from ewatercycle.forcing.forcing_data import ForcingData
+from ewatercycle.models.abstract import AbstractModel
+
+
+@dataclass
+class Solver:
+    """Solver, for current implementations see
+    _`here <https://github.com/wknoben/MARRMoT/tree/master/MARRMoT/Functions/Time%20stepping>`.
+    """
+    name: str = 'createOdeApprox_IE'
+    resnorm_tolerance: float = 0.1
+    resnorm_maxiter: float = 6.0
+
+
+def _generate_work_dir(work_dir: PathLike = None):
+    """
+    Args:
+        work_dir: If work dir is None then create sub-directory in CFG['output_dir']
+    """
+    if work_dir is None:
+        scratch_dir = CFG['output_dir']
+        # TODO this timestamp isnot safe for parallel processing
```

Comment:
This is going to be a problem quickly. Perhaps add milliseconds to kick the can down the road?

Reply:
When I was working with marrmot calibration and parallel processing, I found that this solution does not work because the CPU time would be the same for all cores. We added a random filename generator there.

Reply:
Created https://github.com/eWaterCycle/ewatercycle/issues/67